Python/Product/Analysis/LanguageServer/EditorFiles.cs

```diff
@@ -0,0 +1,218 @@
+                    lock (_pendingChanges) {
+                        _pendingChanges.Add(@params);
+                    }
+                    return;
+                }
+
+                var toVersion = @params.textDocument.version ?? (fromVersion + changes.Length);
+
+                doc.UpdateDocument(part, new DocumentChangeSet(
+                    fromVersion,
+                    toVersion,
+                    changes.Select(c => new DocumentChange {
+                        ReplacedSpan = c.range.GetValueOrDefault(),
+                        WholeBuffer = !c.range.HasValue,
+                        InsertedText = c.text
+                    })
+                ));
+
+                DidChangeTextDocumentParams? next = null;
+                lock (_pendingChanges) {
+                    var notExpired = _pendingChanges
+                        .Where(p => p.textDocument.version.GetValueOrDefault() >= toVersion)
+                        .OrderBy(p => p.textDocument.version.GetValueOrDefault())
+                        .ToArray();
+
+                    _pendingChanges.Clear();
+                    if (notExpired.Any()) {
+                        next = notExpired.First();
+                        _pendingChanges.AddRange(notExpired.Skip(1));
+                    }
+                }
+                if (next.HasValue) {
+                    DidChangeTextDocument(next.Value, null);
+                }
+            } finally {
+                if (enqueueAction != null) {
+                    _server.TraceMessage($"Applied changes to {uri}");
+                    enqueueAction(doc);
+                }
+            }
+        }
+
+        public void UpdateParseDiagnostics(VersionCookie vc, Uri documentUri) {
+            lock (_lock) {
+                var last = _parseBufferDiagnostics;
+                foreach (var kv in vc.GetAllParts(documentUri)) {
+                    var part = _server.ProjectFiles.GetPart(kv.Key);
+                    if (!last.TryGetValue(part, out var lastVersion) || lastVersion.Version < kv.Value.Version || _ignoreDiagnosticsVersion) {
+                        last[part] = kv.Value;
+
+                        PublishDiagnostics(new PublishDiagnosticsEventArgs {
+                            uri = kv.Key,
+                            diagnostics = kv.Value.Diagnostics,
+                            _version = kv.Value.Version
+                        });
+                    }
+                }
+                _ignoreDiagnosticsVersion = false;
+            }
+        }
+
+        public void UpdateAnalysisDiagnostics(IProjectEntry projectEntry, int version) {
+            lock (_lock) {
+                var diags = _server.Analyzer.GetDiagnostics(projectEntry);
+                for (var i = 0; i < diags.Count; i++) {
+                    diags[i].severity = _server.Settings.analysis.GetEffectiveSeverity(diags[i].code, diags[i].severity);
```

Comment:
I think we can cache ` _server.Settings.analysis` at the beginning of the method